modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,237 @@
+	 *
+	 * @param ILO_URL_Metric $new_url_metric New URL metric.
+	 */
+	public function add( ILO_URL_Metric $new_url_metric ) {
+		$url_metrics = $this->flatten();
+		array_unshift( $url_metrics, $new_url_metric );
+
+		$grouped_url_metrics = $this->group_url_metrics_by_breakpoint( $url_metrics );
+
+		// Make sure there is at most $sample_size number of URL metrics for each breakpoint.
+		$grouped_url_metrics = array_map(
+			function ( $breakpoint_url_metrics ) {
+				if ( count( $breakpoint_url_metrics ) > $this->sample_size ) {
+
+					// Sort URL metrics in descending order by timestamp.
+					usort(
+						$breakpoint_url_metrics,
+						static function ( ILO_URL_Metric $a, ILO_URL_Metric $b ): int {
+							return $b->get_timestamp() <=> $a->get_timestamp();
+						}
+					);
+
+					// Only keep the sample size of the newest URL metrics.
+					$breakpoint_url_metrics = array_slice( $breakpoint_url_metrics, 0, $this->sample_size );
+				}
+				return $breakpoint_url_metrics;
+			},
+			$grouped_url_metrics
+		);
+
+		$this->groups = $grouped_url_metrics;
+	}
+
+	/**
+	 * Groups URL metrics by breakpoint.
+	 *
+	 * @since n.e.x.t
+	 * @access private
+	 *
+	 * @param ILO_URL_Metric[] $url_metrics URL metrics.
+	 * @return array<int, ILO_URL_Metric[]> URL metrics grouped by breakpoint. The array keys are the minimum widths for a viewport to lie within
+	 *                                      the breakpoint. The returned array is always one larger than the provided array of breakpoints, since
+	 *                                      the breakpoints reflect the max inclusive boundaries whereas the return value is the groups of page
+	 *                                      metrics with viewports on either side of the breakpoint boundaries.
+	 */
+	private function group_url_metrics_by_breakpoint( array $url_metrics ): array {
+
+		// Convert breakpoint max widths into viewport minimum widths.
+		$minimum_viewport_widths = array_map(
+			static function ( $breakpoint ) {
+				return $breakpoint + 1;
+			},
+			$this->breakpoints
+		);
+
+		$grouped = array_fill_keys( array_merge( array( 0 ), $minimum_viewport_widths ), array() );
+
+		foreach ( $url_metrics as $url_metric ) {
+			$current_minimum_viewport = 0;
+			foreach ( $minimum_viewport_widths as $viewport_minimum_width ) {
+				if ( $url_metric->get_viewport()['width'] >= $viewport_minimum_width ) {
+					$current_minimum_viewport = $viewport_minimum_width;
+				} else {
+					break;
+				}
+			}
+
+			$grouped[ $current_minimum_viewport ][] = $url_metric;
+		}
+		return $grouped;
+	}
+
+	/**
+	 * Determines whether the URL metrics group for a given viewport is lacking URL metrics.
+	 *
+	 * Either the group does not have enough URL metrics for the desired sample size,
+	 * or some of the URL metrics are stale.
+	 *
+	 * @param int $viewport_width Viewport width.
+	 * @return bool Whether group is lacking.
+	 */
+	public function is_group_lacking( int $viewport_width ): bool {
```

Comment:
There are three possible statuses for a URL metrics group:

1. It doesn't have enough URL metrics for the sample size, and/or
2. Some of the URL metrics are stale and need to be refreshed. Or
3. There is a complete number of URL metrics for the desired sample size and they are all fresh (non-stale).

I was thinking of the `lacking` term since it can encapsulate both the first and second statuses. Terms like "full" or "complete" could be misleading since they might describe either status 2 or 3 when the full sample size is present.

Reply:
Thanks for outlining the three states, that makes sense.

I'm not sure about the conclusion though. To me `lacking` is not more clear than `full` or `complete`. I agree `full` may be more limited in meaning, but IMO `complete` would work perfectly well. When you read `lacking`, you may just as much think that the group is lacking items because it doesn't have a sufficient amount, so not really different than the opposite of `complete`.

I think it doesn't really matter though _why_ a group is not `complete`. That's an internal detail that the class itself solves and the caller doesn't need to bother in the first place. That's why from a semantic standpoint, I think `lacking` and `complete` are equally good or bad. However, I think `complete` works better as it is a positive term.

Reply:
Inverted `lacking` to `complete` in da7d2ad0526fddbcb775d1ebca27d7ec67a0fb03.